numpy/lib/tests/test_index_tricks.py

```diff
@@ -566,3 +566,123 @@ def test_ndindex():
     # Make sure 0-sized ndindex works correctly
     x = list(ndindex(*[0]))
     assert_equal(x, [])
+
+
+
+# ndindex test cases for reimplementation of ndindex using itertools 
+# 1.  Empty Shapes and Zero Dimensions
+def test_ndindex_empty_and_zero_dimensions():
+    # Empty shape
+    assert list(np.ndindex()) == [()]
+    assert list(np.ndindex(())) == [()]
+
+    # Single dimension
+    assert list(np.ndindex(4)) == [(0,), (1,), (2,), (3,)]
+
+    # Zero dimensions should produce empty iterator
+    assert list(np.ndindex(0, 3)) == []
+    assert list(np.ndindex(3, 0, 2)) == []
+
+# 2. Non-Integer Dimensions Raise TypeError
+def test_ndindex_non_integer_dimensions():
+    """Test that non-integer dimensions raise TypeError."""
+    with pytest.raises(TypeError):
+        list(np.ndindex(2.5))
+    with pytest.raises(TypeError):
+        list(np.ndindex("2"))
+    with pytest.raises(TypeError):
+        list(np.ndindex([2, 3]))
+    with pytest.raises(TypeError):
+        list(np.ndindex((2.0, 3)))
```

Comment:
minor: could use `pytest.parametrize` for concision and separation of case execution